fem/bilininteg_ext.cpp

```diff
@@ -378,294 +371,376 @@ void PADiffusionApply2D(const int NE,
 {
    const int D1D = T_D1D ? T_D1D : d1d;
    const int Q1D = T_Q1D ? T_Q1D : q1d;
-   MFEM_VERIFY(D1D <= MAX_D1D, "");
-   MFEM_VERIFY(Q1D <= MAX_Q1D, "");
-
-   const DeviceMatrix B(b, Q1D, D1D);
-   const DeviceMatrix G(g, Q1D, D1D);
-   const DeviceMatrix Bt(bt, D1D, Q1D);
-   const DeviceMatrix Gt(gt, D1D, Q1D);
+   const int NBZ = T_NBZ ? T_NBZ : 1;
+   const int MQ1 = T_Q1D ? T_Q1D : MAX_Q1D;
+   const int MD1 = T_D1D ? T_D1D : MAX_D1D;
+   MFEM_VERIFY(MQ1 == MD1, "");
+   MFEM_VERIFY(D1D <= MD1, "");
+   MFEM_VERIFY(Q1D <= MQ1, "");
+   const DeviceMatrix b(_b, Q1D, D1D);
+   const DeviceMatrix g(_g, Q1D, D1D);
+   const DeviceMatrix bt(_bt, D1D, Q1D);
+   const DeviceMatrix gt(_gt, D1D, Q1D);
    const DeviceTensor<3> op(_op, 3, Q1D*Q1D, NE);
    const DeviceTensor<3> x(_x, D1D, D1D, NE);
    DeviceTensor<3> y(_y, D1D, D1D, NE);
-
-   MFEM_FORALL(e, NE,
+   MFEM_FORALL_2D(e, NE, Q1D, Q1D, NBZ,
    {
       const int D1D = T_D1D ? T_D1D : d1d; // nvcc workaround
       const int Q1D = T_Q1D ? T_Q1D : q1d;
-
-      double grad[MAX_Q1D][MAX_Q1D][2];
-      for (int qy = 0; qy < Q1D; ++qy)
+      const int NBZ = T_NBZ ? T_NBZ : 1;
+      const int MQ1 = T_Q1D ? T_Q1D : MAX_Q1D;
+      const int MD1 = T_D1D ? T_D1D : MAX_D1D;
+      MFEM_SHARED double B[MQ1][MD1];
+      MFEM_SHARED double G[MQ1][MD1];
+      MFEM_SHARED double Bt[MD1][MQ1];
+      MFEM_SHARED double Gt[MD1][MQ1];
+      MFEM_SHARED double Xz[NBZ][MD1][MD1];
+      MFEM_SHARED double GD[2][NBZ][MD1][MQ1];
+      MFEM_SHARED double GQ[2][NBZ][MD1][MQ1];
+      double (*X)[MD1] = (double (*)[MD1])(Xz + threadIdx(z));
+      double (*DQ0)[MD1] = (double (*)[MD1])(GD[0] + threadIdx(z));
+      double (*DQ1)[MD1] = (double (*)[MD1])(GD[1] + threadIdx(z));
+      double (*QQ0)[MD1] = (double (*)[MD1])(GQ[0] + threadIdx(z));
+      double (*QQ1)[MD1] = (double (*)[MD1])(GQ[1] + threadIdx(z));
+      for (int dy = threadIdx(y); dy < D1D; dy += blockDim(y))
       {
-         for (int qx = 0; qx < Q1D; ++qx)
+         for (int dx = threadIdx(x); dx < D1D; dx += blockDim(x))
          {
-            grad[qy][qx][0] = 0.0;
-            grad[qy][qx][1] = 0.0;
+            X[dy][dx] = x(dx,dy,e);
          }
       }
-      for (int dy = 0; dy < D1D; ++dy)
+      if (threadIdx(z) == 0)
       {
-         double gradX[MAX_Q1D][2];
-         for (int qx = 0; qx < Q1D; ++qx)
+         for (int dx = threadIdx(y); dx < D1D; dx += blockDim(y))
          {
-            gradX[qx][0] = 0.0;
-            gradX[qx][1] = 0.0;
+            for (int qx = threadIdx(x); qx < Q1D; qx += blockDim(x))
+            {
+               B[qx][dx] = b(qx,dx);
+               G[qx][dx] = g(qx,dx);
+               Bt[dx][qx] = bt(dx,qx);
```

Comment:
You can save a little bit of bandwidth by pulling the results for Bt/Gt out of b/g instead of bt/gt.

Reply:
Actually in the shared memory case it probably makes sense to not store the transpose versions since we aren't concerned with coalesced memory reads out of shared memory, but we are concerned about using up all the shared memory.